### DISCUSSION TITLE
no method named `split` found for `tokio_util::codec::Framed<tokio::net::TcpStream, ...>` after upgrade to 0.3

### DISCUSSION THREAD

**Discussion by martinetd** [tokio-rs/tokio#3081] [Q&A] [answered]
I'm trying to upgrade to tokio 0.3

At first I ran into the same issue as https://github.com/tokio-rs/tokio/discussions/2984 because I hadn't upgraded tokio-util at the same time, but I think that's different now I've also updated it. I get errors like the following:
```
error[E0599]: no method named `split` found for struct `tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>` in the current scope
   --> src/main.rs:865:39
    |
865 |       let (mut writer, reader) = stream.split();
    |                                         ^^^^^ method not found in `tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>`
    | 
   ::: /home/asmadeus/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-util-0.5.0/src/codec/framed.rs:18:1
    |
18  | / pin_project! {
19  | |     /// A unified [`Stream`] and [`Sink`] interface to an underlying I/O object, using
20  | |     /// the `Encoder` and `Decoder` traits to encode and decode frames.
21  | |     ///
...   |
32  | |     }
33  | | }
    | | -
    | | |
    | | doesn't satisfy `_: futures::AsyncReadExt`
    | | doesn't satisfy `_: futures::AsyncRead`
    | |_doesn't satisfy `_: futures::StreamExt`
    |   doesn't satisfy `_: futures::Stream`
    |
    = note: the method `split` exists but the following trait bounds were not satisfied:
            `tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>: futures::AsyncRead`
            which is required by `tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>: futures::AsyncReadExt`
            `tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>: futures::Stream`
            which is required by `tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>: futures::StreamExt`
            `&tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>: futures::AsyncRead`
            which is required by `&tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>: futures::AsyncReadExt`
            `&tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>: futures::Stream`
            which is required by `&tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>: futures::StreamExt`
            `&mut tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>: futures::AsyncRead`
            which is required by `&mut tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>: futures::AsyncReadExt`
            `&mut tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>: futures::Stream`
            which is required by `&mut tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>: futures::StreamExt`
```

But the doc hasn't changed on this point so I'm not sure what to do.

From what I understand tokio's TcpStream implements tokio's AsyncRead/Write and not futures' so maybe a mix between the two? I now see there also is another split in tokio with ReadHalf/WriteHalf which have tokio's AsyncRead/Write and might be more approprite but I can't fiddle things properly to get this right.
I'm sure this has to be documented somewhere but there are a lot of outdated examples out there and I'm just about completely lost, to be honest I think it was a miracle that it worked in the first place previously :D

You can try on https://github.com/martinetd/matrirc/tree/tokio-0.3 (previous commit works with tokio 0.2/tokio-util 0.3)

Thanks!

**Comment by Darksonn** [comment, marked as answer]
Rust couldn't decide which split you wanted, so I changed the line to the following to disambiguate it:
```rust
let (mut writer, reader) = futures::StreamExt::split(stream);
```
this gives me:
```
error[E0277]: the trait bound `irc::irc_proto::IrcCodec: tokio_util::codec::Decoder` is not satisfied
    --> src/main.rs:865:58
     |
865  |     let (mut writer, reader) = futures::StreamExt::split(stream);
     |                                                          ^^^^^^ the trait `tokio_util::codec::Decoder` is not implemented for `irc::irc_proto::IrcCodec`
     | 
    ::: /home/alice/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.7/src/stream/stream/mod.rs:1277:14
     |
1277 |     fn split<Item>(self) -> (SplitSink<Self, Item>, SplitStream<Self>)
     |              ---- required by this bound in `futures::StreamExt::split`
     |
     = note: required because of the requirements on the impl of `futures::Stream` for `tokio_util::codec::Framed<tokio::net::TcpStream, irc::irc_proto::IrcCodec>`
```
So it appears that the `IrcCodec` doesn't implement the `Decoder` trait. Perhaps this is a `tokio-util` version mismatch? If so, you can probably write a wrapper struct like this:
```rust
struct MyIrcCodec {
    inner: IrcCodec,
}
```
then implement the new `Decoder` trait by just forwarding the call to the method from the old `Decoder` trait on the inner type.

**Reply by martinetd** [reply]
> then implement the new Decoder trait by just forwarding the call to the method from the old Decoder trait on the inner type.

Thanks again for the help (ages ago); this post has been helpful again!

FWIW I had the same problem again with tokio-util 0.6 -> 0.7 and I tried to wrap it, but couldn't manage to call the inner decoder/encoder implem.

You can try it here, in its broken state:
https://github.com/martinetd/matrirc/commits/tokio_util_incompat

Calling self.inner.decode naively just says "hey it's implemented in tokio-util 0.6!":
```
error[E0599]: no method named `decode` found for struct `IrcCodec` in the current scope
   --> src/ircd/codec.rs:34:20
    |
34  |         self.inner.decode(src)
    |                    ^^^^^^ method not found in `IrcCodec`
    |
   ::: /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-util-0.6.10/src/codec/decoder.rs:124:8
    |
124 |     fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error>;
    |        ------ the method is available for `IrcCodec` here
    |
    = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
    |
5   | use tokio_util::codec::decoder::Decoder;
    |
```

But calling Decoder::decode directly as a debug measure fails as I only have the tokio-util 0.7 in scope:
```
error[E0277]: the trait bound `IrcCodec: Decoder` is not satisfied
  --> src/ircd/codec.rs:35:9
   |
35 |         Decoder::decode(&mut self.inner, src)
   |         ^^^^^^^^^^^^^^^ the trait `Decoder` is not implemented for `IrcCodec`
   |
   = help: the following other types implement trait `Decoder`:
             AnyDelimiterCodec
             BytesCodec
             Codec
             LengthDelimitedCodec
             LinesCodec
```

Turns out the irc crate already updated to 0.7 in an unreleased commit so I just bumped the commit instead and there's no problem anymore, but this is really painful so would have wanted to wrap it up if possible so I could forget about this in the future.

As far as I understand, if the irc::proto crate would re-export tokio_util then it'd be possible to use `irc::proto::tokio_util::codec::Decoder` and get the old 0.6 version, but since they're not doing this I'm not aware of any way of getting this right -- is that correct?

Cheers.

**Comment by martinetd** [comment]
> Perhaps this is a `tokio-util` version mismatch?

Right, the irc crate is still using `tokio-util` 0.3 -- the "trick" of specifying the correct split helped thanks.

> then implement the new `Decoder` trait by just forwarding the call to the method from the old Decoder trait on the inner type.

I'll try to get the dep updated instead, although doing that would probably be good for the sake of learning how...

Thanks!
